File: server/matchingEngine.ts
/**
 * MatchingEngine — Ported from com.icare.service.MatchingEngine (Java)
 *
 * Haversine distance calculation and hospital matching/scoring.
 */

import { Hospital, getHospitals } from "./dbService.js";

// ─── Distance Calculator (ported from DistanceCalculator.java) ───

const EARTH_RADIUS_KM = 6371.0;

export function haversineDistance(lat1: number, lng1: number, lat2: number, lng2: number): number {
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const latDelta = toRad(lat2 - lat1);
  const lonDelta = toRad(lng2 - lng1);
  const a =
    Math.sin(latDelta / 2) * Math.sin(latDelta / 2) +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
    Math.sin(lonDelta / 2) * Math.sin(lonDelta / 2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return EARTH_RADIUS_KM * c;
}

// ─── Types ───────────────────────────────────────────────────────

export interface HospitalMatch {
  hospital: Hospital;
  distanceKm: number;
  suitabilityScore: number;
}

export interface UserProfile {
  latitude: number;
  longitude: number;
  chronicConditions?: string[];
  allergies?: string[];
  medications?: string[];
  bloodType?: string;
}

// ─── Condition → Service Mapping ─────────────────────────────────

const CONDITION_TO_SERVICE: Record<string, string> = {
  kidney: "Dialysis",
  diabetes: "Endocrinology",
  heart: "Cardiology",
  maternal: "Maternity",
  pregnancy: "Maternity",
  pediatrics: "Pediatrics",
  fever: "Internal Medicine",
  respiratory: "Pulmonology",
  wound: "Emergency",
  stroke: "Neurology",
};

// ─── Scoring ─────────────────────────────────────────────────────

function calculateSuitabilityScore(distanceKm: number, matchingKeywords: number): number {
  const distanceScore = Math.max(0.0, 100.0 - distanceKm * 3.0);
  const relevanceBonus = matchingKeywords * 20.0;
  return Math.round((distanceScore + relevanceBonus) * 100.0) / 100.0;
}

function countMatchingServices(hospital: Hospital, needs: Set<string>): number {
  if (!hospital.services?.length || needs.size === 0) return 0;

  let count = 0;
  for (const service of hospital.services) {
    if (!service) continue;
    const normalizedService = service.toLowerCase();
    for (const need of needs) {
      if (normalizedService.includes(need) || need.includes(normalizedService)) {
        count++;
        break;
      }
      const mappedService = CONDITION_TO_SERVICE[need];
      if (mappedService && normalizedService.includes(mappedService.toLowerCase())) {
        count++;
        break;
      }
    }
  }
  return count;
}

function buildSearchTerms(profile: UserProfile): Set<string> {
  const terms = new Set<string>();
  profile.chronicConditions?.forEach(c => { if (c?.trim()) terms.add(c.trim().toLowerCase()); });
  profile.allergies?.forEach(a => { if (a?.trim()) terms.add(a.trim().toLowerCase()); });
  profile.medications?.forEach(m => { if (m?.trim()) terms.add(m.trim().toLowerCase()); });
  if (profile.bloodType?.trim()) terms.add(profile.bloodType.trim().toUpperCase());
  return terms;
}

// ─── Public API ──────────────────────────────────────────────────

export function findNearbyHospitals(
  latitude: number,
  longitude: number,
  radiusKm: number = 10,
  limit: number = 6
): HospitalMatch[] {
  const hospitals = getHospitals();

  return hospitals
    .map(hospital => ({
      hospital,
      distanceKm: haversineDistance(latitude, longitude, hospital.latitude, hospital.longitude),
      suitabilityScore: 0,
    }))
    .filter(match => match.distanceKm <= radiusKm)
    .sort((a, b) => a.distanceKm - b.distanceKm)
    .slice(0, limit);
}

export function findMatchingHospitals(profile: UserProfile): HospitalMatch[] {
  const hospitals = getHospitals();
  if (!hospitals.length) return [];

  const normalizedNeeds = buildSearchTerms(profile);

  return hospitals
    .map(hospital => {
      const distanceKm = haversineDistance(profile.latitude, profile.longitude, hospital.latitude, hospital.longitude);
      const matchingKeywords = countMatchingServices(hospital, normalizedNeeds);
      const score = calculateSuitabilityScore(distanceKm, matchingKeywords);
      return { hospital, distanceKm, suitabilityScore: score };
    })
    .filter(match => match.suitabilityScore >= 0)
    .sort((a, b) => b.suitabilityScore - a.suitabilityScore);
}

File: backend/src/main/java/com/icare/service/MatchingEngine.java
package com.icare.service;

import com.icare.model.Hospital;
import com.icare.model.UserProfile;
import com.icare.util.DistanceCalculator;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class MatchingEngine {

    private final ExcelService excelService;

    private static final Map<String, String> CONDITION_TO_SERVICE = Map.ofEntries(
            Map.entry("kidney", "Dialysis"),
            Map.entry("diabetes", "Endocrinology"),
            Map.entry("heart", "Cardiology"),
            Map.entry("maternal", "Maternity"),
            Map.entry("pregnancy", "Maternity"),
            Map.entry("pediatrics", "Pediatrics"),
            Map.entry("fever", "Internal Medicine"),
            Map.entry("respiratory", "Pulmonology"),
            Map.entry("wound", "Emergency"),
            Map.entry("stroke", "Neurology")
    );

    public MatchingEngine(ExcelService excelService) {
        this.excelService = excelService;
    }

    public List<HospitalMatch> findMatchingHospitals(UserProfile profile) {
        List<Hospital> hospitals = excelService.getHospitals();
        if (profile == null || hospitals.isEmpty()) {
            return new ArrayList<>();
        }

        Set<String> normalizedNeeds = buildSearchTerms(profile);
        double userLatitude = profile.getLatitude();
        double userLongitude = profile.getLongitude();

        return hospitals.stream()
                .map(hospital -> buildMatch(hospital, normalizedNeeds, userLatitude, userLongitude))
                .filter(match -> match.getSuitabilityScore() >= 0)
                .sorted(Comparator.comparingDouble(HospitalMatch::getSuitabilityScore).reversed())
                .collect(Collectors.toList());
    }

    public List<HospitalMatch> findNearbyHospitals(double latitude, double longitude, int radiusKm, int limit) {
        List<Hospital> hospitals = excelService.getHospitals();
        if (hospitals.isEmpty()) {
            return new ArrayList<>();
        }

        return hospitals.stream()
                .map(hospital -> {
                    double distanceKm = DistanceCalculator.haversineDistance(latitude, longitude, hospital.getLatitude(), hospital.getLongitude());
                    return new HospitalMatch(hospital, distanceKm, 0.0);
                })
                .filter(match -> match.getDistanceKm() <= radiusKm)
                .sorted(Comparator.comparingDouble(HospitalMatch::getDistanceKm))
                .limit(limit)
                .collect(Collectors.toList());
    }

    private Set<String> buildSearchTerms(UserProfile profile) {
        Set<String> terms = new HashSet<>();
        if (profile.getChronicConditions() != null) {
            for (String condition : profile.getChronicConditions()) {
                if (condition != null && !condition.isBlank()) {
                    terms.add(condition.trim().toLowerCase());
                }
            }
        }
        if (profile.getAllergies() != null) {
            for (String allergy : profile.getAllergies()) {
                if (allergy != null && !allergy.isBlank()) {
                    terms.add(allergy.trim().toLowerCase());
                }
            }
        }
        if (profile.getMedications() != null) {
            for (String medication : profile.getMedications()) {
                if (medication != null && !medication.isBlank()) {
                    terms.add(medication.trim().toLowerCase());
                }
            }
        }
        if (profile.getBloodType() != null && !profile.getBloodType().isBlank()) {
            terms.add(profile.getBloodType().trim().toUpperCase());
        }
        return terms;
    }

    private HospitalMatch buildMatch(Hospital hospital, Set<String> normalizedNeeds, double userLatitude, double userLongitude) {
        double distanceKm = DistanceCalculator.haversineDistance(userLatitude, userLongitude, hospital.getLatitude(), hospital.getLongitude());
        int matchingKeywords = countMatchingServices(hospital, normalizedNeeds);
        double score = calculateSuitabilityScore(distanceKm, matchingKeywords);
        return new HospitalMatch(hospital, distanceKm, score);
    }

    private int countMatchingServices(Hospital hospital, Set<String> needs) {
        if (hospital.getServices() == null || hospital.getServices().isEmpty() || needs.isEmpty()) {
            return 0;
        }
        int count = 0;
        for (String service : hospital.getServices()) {
            if (service == null) {
                continue;
            }
            String normalizedService = service.toLowerCase();
            for (String need : needs) {
                if (normalizedService.contains(need) || need.contains(normalizedService)) {
                    count++;
                    break;
                }
                String mappedService = CONDITION_TO_SERVICE.get(need);
                if (mappedService != null && normalizedService.contains(mappedService.toLowerCase())) {
                    count++;
                    break;
                }
            }
        }
        return count;
    }

    private double calculateSuitabilityScore(double distanceKm, int matchingKeywords) {
        double distanceScore = Math.max(0.0, 100.0 - (distanceKm * 3.0));
        double relevanceBonus = matchingKeywords * 20.0;
        return Math.round((distanceScore + relevanceBonus) * 100.0) / 100.0;
    }

    public static class HospitalMatch {
        private final Hospital hospital;
        private final double distanceKm;
        private final double suitabilityScore;

        public HospitalMatch(Hospital hospital, double distanceKm, double suitabilityScore) {
            this.hospital = hospital;
            this.distanceKm = distanceKm;
            this.suitabilityScore = suitabilityScore;
        }

        public Hospital getHospital() {
            return hospital;
        }

        public double getDistanceKm() {
            return distanceKm;
        }

        public double getSuitabilityScore() {
            return suitabilityScore;
        }
    }
}
